pkg/resources/poddistribution/poddistribution.go

```diff
@@ -0,0 +1,386 @@
+	// ZoneLabel is the label that specifies the zone where a node is
+	ZoneLabel = "topology.kubernetes.io/zone"
+	// Annotation counter on the pods controller, dc, rs, ss
+	PodBalanceAttempts = "pod-balance-attempts"
+	maxBalanceAttempts = 3
+	name               = "name"
+	ns                 = "ns"
+	kind               = "kind"
+)
+
+type KindNameSpaceName struct {
+	*k8sTypes.NamespacedName
+	Kind string
+}
+
+func (knn KindNameSpaceName) String() string {
+	return fmt.Sprintf("%s/%s/%s", knn.Kind, knn.Namespace, knn.Name)
+}
+
+// Check the PodBalanceAttempts, ensure less than maxBalanceAttempts
+func verifyBalanceCount(ctx context.Context, client k8sclient.Client, obj metav1.Object) bool {
+	ant := obj.GetAnnotations()
+	// If there are no annotations then 0 balance attempts have been made
+	if ant == nil {
+		return true
+	}
+	if val, ok := ant[PodBalanceAttempts]; ok {
+		i, err := strconv.Atoi(val)
+		if err != nil {
+			logrus.Errorf("Error converting string annotations %s", ant[PodBalanceAttempts])
+			return false
+		} else {
+			if i >= maxBalanceAttempts {
+				logrus.Warningf("Reached max balance attempts for %s on %s", obj.GetName(), obj.GetNamespace())
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func getObject(ctx context.Context, client k8sclient.Client, objDtls map[string]string) (metav1.Object, error) {
+	if objDtls[kind] == "DeploymentConfig" {
+		return getDeploymentConfig(ctx, client, objDtls)
+	}
+	if objDtls[kind] == "ReplicaSet" {
+		return getReplicaSet(ctx, client, objDtls)
+	}
+	if objDtls[kind] == "StatefulSet" {
+		return getStatefulSet(ctx, client, objDtls)
+	}
+	logrus.Errorf("Unexpected obj kind %s", objDtls[kind])
+	return nil, nil
+}
+
+func getDeploymentConfig(ctx context.Context, client k8sclient.Client, objDtls map[string]string) (metav1.Object, error) {
+	deploymentConfig := &appsv1.DeploymentConfig{}
+	err := client.Get(ctx, k8sTypes.NamespacedName{Name: objDtls[name], Namespace: objDtls[ns]}, deploymentConfig)
+	if err != nil {
+		logrus.Errorf("Error getting DeploymentConfig %v", err)
+		return nil, err
+	}
+	return deploymentConfig, nil
+}
+
+func getReplicaSet(ctx context.Context, client k8sclient.Client, objDtls map[string]string) (metav1.Object, error) {
+	rs := &k8appsv1.ReplicaSet{}
+	err := client.Get(ctx, k8sTypes.NamespacedName{Name: objDtls[name], Namespace: objDtls[ns]}, rs)
+	if err != nil {
+		logrus.Errorf("Error getting ReplicaSet %v", err)
+		return nil, err
+	}
+	return rs, nil
+}
+
+func getStatefulSet(ctx context.Context, client k8sclient.Client, objDtls map[string]string) (metav1.Object, error) {
+	ss := &k8appsv1.StatefulSet{}
+	err := client.Get(ctx, k8sTypes.NamespacedName{Name: objDtls[name], Namespace: objDtls[ns]}, ss)
+	if err != nil {
+		logrus.Errorf("Error getting StatefulSet %v", err)
+		return nil, err
+	}
+	return ss, nil
+}
+
+func parseKnn(knn string) map[string]string {
+	objDtls := map[string]string{}
+	s := strings.Split(knn, "/")
+	objDtls[kind] = s[0]
```

Comment:
is it possible this will be malformed and we end up with an indexOutOfBound error?

Reply:
Will add check here